skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt

```diff
@@ -0,0 +1,122 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold
+
+import com.google.common.annotations.VisibleForTesting
+import com.intellij.openapi.components.ServiceManager
+import java.io.File
+import java.nio.file.Path
+import java.nio.file.Paths
+
+/**
+ * Interface for Skaffold execution service. This service builds and launches Skaffold process from
+ * the given set of standard flags and settings. [DefaultSkaffoldExecutorService] assumes Skaffold
+ * to be already installed and available.
+ */
+interface SkaffoldExecutorService {
+    companion object {
+        val instance
+            get() = ServiceManager.getService(SkaffoldExecutorService::class.java)!!
+    }
+
+    /**
+     * Creates Skaffold command line from the given settings and returns resulting launched
+     * Skaffold system process.
+     *
+     * @param settings Settings with flags supported by Skaffold.
+     * @return [SkaffoldProcess] with resulting process and the command line used to launch it.
+     */
+    fun executeSkaffold(settings: SkaffoldExecutorSettings): SkaffoldProcess
+
+    /**
+     * Execution mode for Skaffold, single run, continuous development, etc.
+     */
+    enum class SkaffoldExecutionMode(val modeFlag: String) {
+        SINGLE_RUN("run"),
+        DEV("dev")
+    }
+
+    /**
+     * Set of settings to control Skaffold execution, including flags and execution mode.
+     *
+     * @property executionMode Mandatory execution mode for Skaffold, see [SkaffoldExecutionMode].
+     * @property skaffoldConfigurationFilePath Optional, location of the Skaffold YAML
+     *           configuration file. If not provided, default `skaffold.yaml` used.
+     * @property workingDirectory Optional, working directory where Skaffold needs to be launched.
+     *           This is usually set to project working directory.
+     */
+    data class SkaffoldExecutorSettings(
+        val executionMode: SkaffoldExecutionMode,
+        val skaffoldConfigurationFilePath: String? = null,
+        var workingDirectory: File? = null
+    )
+
+    /**
+     * Data object with launched Skaffold process and its command line.
+     *
+     * @property process System process for Skaffold.
+     * @property commandLine Command line used to launch the process.
+     */
+    data class SkaffoldProcess(val process: Process, val commandLine: String)
+}
+
+/**
+ * Abstract implementation of [SkaffoldExecutorService], constructs Skaffold command line and
+ * build resulting process. Path to Skaffold executable must be set by subclasses.
+ */
+abstract class AbstractSkaffoldExecutorService : SkaffoldExecutorService {
```

Comment:
Kotlin interfaces can contain method implementations. Given that, is the abstract class still needed or can you push things up to the interface?

Reply:
It feels like interface declares just an intention, and all needed data structures like `SkaffoldExecutorSettings`, so it's pretty full of information as it is, and abstract class describes the actual execution details. We can move that to interface, but it may make things a bit cluttered. wdyt?

Reply:
I agree, simplified a bit. PTAL